app/services/emailAuth.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  sendSignInLinkToEmail,
+  signInWithEmailLink,
+  signOut,
+} from 'firebase/auth';
+import { auth, actionCodeSettings } from '../config/firebase';
+import userDataSanitizationSchema from '../schemas/userDataSanitizationSchema';
+import { storeUserDataIfNew } from './firestore';
+
+/**
+ * Sends authentication link to user email and saves user email address to local storage.
+ * @param {string} email - user's email address
+ */
+export const sendEmailLink = (email: string) => {
+  sendSignInLinkToEmail(auth, email, actionCodeSettings)
+    .then(() => {
+      window.localStorage.setItem('emailForSignIn', email);
+    })
+    .catch((error) => {
+      console.error(
+        'Error occurred while sending email authentication link. Please try again.'
+      );
+    });
+};
+
+/**
+ * Authenticates and signs in user following redirect. Then saves their form data to database.
+ * @param {email} email - User's email address. Must be passed or user will be asked to provide it.
+ */
+export const authenticateAndSaveUserDataFromEmailRedirect = (email: string) => {
+  signInWithEmailLink(auth, email, window.location.href)
+    .then(async (result) => {
```

Comment:
Looks like this function mixes async/await with promise chains. To improve readability and consistency, consider refactoring this function to use only async/await throughout. Below is an example: 

```
export const authenticateAndSaveUserDataFromEmailRedirect = async (email: string) => {
  try {
    const result = await signInWithEmailLink(auth, email, window.location.href);
    const localStorageData = {…}
…
  } catch (error) {
    console.error('...’);
  }
};
```